File: auction-app-ui/src/services/productService.ts
import { Product } from 'models/product';
import agent from 'lib/agent';
import { CreateProductDataRequest } from 'models/request/create/createProductDataRequest';
import { CreatePaymentRequest } from 'models/request/create/createPaymentRequest';

const BASE_URL = '/products';

const productsService = {
  getProducts: (params: {}) => agent.get<any>(BASE_URL, params),
  getSingleProduct: (id: string) => agent.get<Product>(`${BASE_URL}/${id}`),
  getRandomProduct: () => agent.get<any>(`${BASE_URL}/random`),
  search: (queryParam: string) => agent.get<any>(`${BASE_URL}/search?criteria=${queryParam}`),
  getUserProducts: (userId: string) => agent.get<Product[]>(`${BASE_URL}/user/${userId}`),
  addProduct: (createProductDataRequest: CreateProductDataRequest) => agent.post<Product>(BASE_URL, createProductDataRequest),
  getRelatedProducts: (params: {}) => agent.get<any>(`${BASE_URL}/related`, params),
  pay: (createPaymentRequest: CreatePaymentRequest) => agent.post<boolean>(`${BASE_URL}/pay`, createPaymentRequest),
};

export default productsService;

File: auction-app-backend/src/main/java/com/internship/auctionapp/services/product/ProductService.java
package com.internship.auctionapp.services.product;

import com.internship.auctionapp.models.Payment;
import com.internship.auctionapp.models.Product;
import com.internship.auctionapp.entities.ProductEntity;
import com.internship.auctionapp.requests.CreatePaymentRequest;
import com.internship.auctionapp.requests.CreateProductDataRequest;

import com.internship.auctionapp.requests.ProductEventRequest;
import com.internship.auctionapp.requests.SearchProductRequest;
import org.springframework.data.domain.Page;

import java.util.List;
import java.util.UUID;

public interface ProductService {
    Page<Product> getProducts(SearchProductRequest searchProductRequest);

    Product addProduct(CreateProductDataRequest createProductDataRequest);

    Product getSingleProduct(UUID id);

    Product updateProduct(UUID id, ProductEntity product);

    void deleteProduct(UUID id);

    Page<Product> getRandomProduct();

    Page<Product> getProductsByCriteria(String criteria);

    List<Product> getUserProducts(UUID userId);

    Page<Product> getRelatedProducts(UUID categoryId, UUID productId);

    Payment purchase(String username, CreatePaymentRequest createPaymentRequest);

    void createNotificationsAfterProductExpires();

    void emitEventOnProductBid(ProductEventRequest productEventRequest, UUID productId);
}
